### PR TITLE
(PUP-3379) Bring back retry action

### PR DESCRIPTION
**Pull request by joshcooper**
Commit 4f1ff7d removed the inventory service code, and since that was
the only consumer of Puppet::Util::RetryAction, it was removed too.

But we have other cases where that behavior would useful, so this
commit adds it back, with some improvements.

### CODE REVIEW

**Comment by joshcooper** [comment]
ping @MikaelSmith

**Comment by MikaelSmith** [comment]
Rob mentioned naming it retry_action for naming consistency (https://github.com/puppetlabs/puppet/pull/3107/files#r18297065). That seems reasonable to do here.

File: spec/unit/util/retryaction_spec.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      Puppet::Util::RetryAction.retry_action( :retries => 0 ) do
+        raise NoMemoryError, "OOM"
+      end
+    end.to raise_error(NoMemoryError, /OOM/)
+  end
+
+  it 'should retry on any exception if no acceptable exceptions given' do
+    Puppet::Util::RetryAction.expects(:sleep).with( (((2 ** 1) -1) * 0.1) )
+    Puppet::Util::RetryAction.expects(:sleep).with( (((2 ** 2) -1) * 0.1) )
+
```
Comment (MikaelSmith) [inline spec/unit/util/retryaction_spec.rb]: Should there be something saying we don't expect :sleep called with other arguments?

Comment (joshcooper) [inline spec/unit/util/retryaction_spec.rb]: If sleep is called with different arguments, then the above expectation will fail. Or do you mean we should verify that sleep is called with an exponentially increasing delay?

**Comment by joshcooper** [comment]
@MikaelSmith @kylog I've added a commit to rename `retryaction` to `retry_action` per https://github.com/puppetlabs/puppet/pull/3375#issuecomment-69370505

File: spec/unit/util/retryaction_spec.rb (hunk 1 above)
Comment (MikaelSmith) [inline spec/unit/util/retryaction_spec.rb]: Ok, I wasn't sure if it would fail on different arguments.

**Comment by joshcooper** [comment]
@MikaelSmith I think everything has been addressed exception for you question in https://github.com/puppetlabs/puppet/pull/3375#discussion_r22745783. I'm not sure if I answered your question there.

**Comment by MikaelSmith** [comment]
Looks good to me :+1:

**Pull request merged**